pylint/checkers/strings.py

```diff
@@ -787,3 +871,66 @@ def str_eval(token):
     if token[0:3] in ('"""', "'''"):
         return token[3:-3]
     return token[1:-1]
+
+
+def is_long_string(token: str) -> bool:
+    """Is this token a "longstring" (is it well-formed and triple-quoted)?
+
+    Long strings are triple-quoted as defined in
+    https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals
+
+    Args:
+        token: The token to be parsed.
+
+    Returns:
+        A boolean representing whether or not this token matches a longstring
+        regex.
+    """
+    single_quoted_regex = "(%s)?'''.*'''\Z" % "|".join(_STRING_AND_BYTES_PREFIXES)
+    double_quoted_regex = '(%s)?""".*"""\Z' % "|".join(_STRING_AND_BYTES_PREFIXES)
+
+    return re.match(single_quoted_regex, token, re.DOTALL) or re.match(
+        double_quoted_regex, token, re.DOTALL
+    )
+
+
+def get_quote_delimiter(token: str) -> Optional[str]:
+    """Returns the quote character used to delimit this token string.
+
+    This function does little checking for whether the token is a well-formed
+    string.
+
+    Args:
+        token: The token to be parsed.
+
+    Returns:
+        A string containing solely the first quote delimiter character in the
+        passed string.  If no quote delimiter characters are present, `None` is
+        returned instead.
+    """
+    match = re.match(
+        "(%s)?(\"|')" % "|".join(_STRING_AND_BYTES_PREFIXES), token, re.DOTALL
```

Comment:
consider compiling all these regular expressions as globals, just after you define `_STRING_AND_BYTES_PREFIXES`

Reply:
I'm torn, here.

I expect our end decision to be purely about readability.  I haven't run any benchmarks, but per [the docs](https://docs.python.org/3/library/re.html#re.compile):

> the module-level matching functions are cached, so programs that use only a few regular expressions at a time needn’t worry about compiling regular expressions.

I generally prefer pulling strings and other magic numbers out from functions to module-level.  I like how it encourages me to refer to them by name rather by literal and give them more meaning in so doing.  I like how it separates logic and instance details, helping a reader focus on both independently.

But here, I find myself not wanting to do that.  The instance details seem more tightly coupled with the logic.  When I read `match.group(2)`, I need to look at the regex in question to understand what it refers to.  That's easier when the regex is right there than if it were at the top of the file.  The long_string regexes aren't quite so bad, in this respect -- if, even after reading this, you think it reads better to have those up top, I'd happily move them.

I couldn't find any guidance in the style guide, but did find this from Nathaniel in g/python-style:

> It is never wrong to move a code element that can be at module scope from function scope to module scope. It is only sometimes wrong to refrain from doing so. Some authors like the locality of having constants close to their one use-site; other authors like the way that migrating to module scope everything that can be so migrated shrinks and simplifies their function implementations.

What do you think about, yourself, when making this sort of judgment call?

Reply:
i didn't know the runtime cached calls to `re.compile`! the existing code should be fine, then. you could do the join as a global variable if you like, so your code would be `re.match("(%s)?..." % _STRING_AND_BYTES_PREFIXES)` without the need to repeat the string join everywhere, more as a matter of the code looking uncluttered than performance. (i also think a shorter name like `_STRING_PREFIXES` or even just `_PREFIXES` given that the file is `strings.py` would be fine, though that's just a matter of personal taste)

Reply:
I like the implicit context with the name `_PREFIXES` a lot.  I've made that change.

You're right that it would unclutter the code, but I've held off on creating a global representing the string in `"|".join(_PREFIXES)` because that string seems like a combination of technical implementation and instance.  I can't think of a clear and succinct meaning it has as a string, separated from its use in the regex.  In this case, I prefer the clutter because of that clearer meaning.